### PR TITLE
fix(grpo): clip the KL log-ratio before exp to avoid inf overflow

### LINKED ISSUE #3015: Numerical stability in KL's exp function
I experienced some numerical issues here
https://github.com/huggingface/trl/blob/e3244d2d096ff1e2e248c931d06d39e165e20623/trl/trainer/grpo_trainer.py#L943

My `ref_per_token_logps` and `per_token_logps` differ more during training, leading to a large value in `exp` and getting `inf` on the step.

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
The K3 KL estimator can overflow when policy and reference log probabilities drift far apart. I added optional log-ratio clipping that keeps the exponential finite, preserves the gradient through clipped values, and rejects invalid clip settings. CPU tests exercise the trainer's loss path for overflow, unchanged defaults, gradient direction, and validation, and lint and format checks pass.
Resolves #3015.

---

> [!NOTE]
> **Medium Risk**
> Touches core GRPO loss math when users opt in via `kl_log_ratio_clip`; default-off path is intended to stay bit-identical, but incorrect clip choice could bias KL gradients during RL training.
> 
> **Overview**
> Fixes **#3015** by adding optional **`GRPOConfig.kl_log_ratio_clip`** (default `None`, unchanged behavior when unset). When `beta > 0`, the K3 KL term now upper-bounds `ref − policy` log-ratios with a **straight-through** clamp before `exp`, so large policy drift no longer blows up the loss to `inf`.
> 
> **`GRPOConfig`** validates the clip (positive, finite) and rejects values whose own `exp` would overflow the log-prob dtype. **`GRPOTrainer`** applies the same logic in `_compute_loss` and **raises `NotImplementedError`** if `use_liger_kernel=True` with a non-`None` clip, since the fused Liger path cannot honor it.
> 
> The identical K3 block is mirrored in **`gmpo_trainer`** and **`gspo_token`** experimental trainers.
> 
> **Tests** add regression coverage: overflow vs finite loss, invalid clip values, one-sided clipping, gradient sign toward the reference (with/without bias correction), non-negative KL, config defaults, and Liger incompatibility.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0aaea03f2fa449bc7a91f1973e7940da11da65da. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Thanks, I'm not strongly opposed, but afaik, no-one is using a ref model regularization anymore. So it'm not sure if adding a new parameter *again* is really worth it

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -3090,9 +3090,12 @@ def _compute_loss(self, model, inputs):
...
         if self.beta != 0.0:
             ref_per_token_logps = inputs["ref_per_token_logps"]
-            per_token_kl = (
-                torch.exp(ref_per_token_logps - per_token_logps) - (ref_per_token_logps - per_token_logps) - 1
-            )
+            log_ratio = ref_per_token_logps - per_token_logps
+            # Clip the log-ratio before the exponential so the K3 estimator stays finite when the policy and
+            # reference distributions drift far apart (issue #3015). No-op when kl_log_ratio_clip is None.
+            if self.args.kl_log_ratio_clip is not None:
+                log_ratio = log_ratio.clamp(-self.args.kl_log_ratio_clip, self.args.kl_log_ratio_clip)
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py]: Confirmed real and fixed in 707808f7 (thread auto-resolved on push; leaving a note for traceability). The KL guard reused the name `log_ratio` inside the `beta != 0` block, shadowing the importance-sampling log-ratio (`per_token_logps - old_per_token_logps`) consumed later via `log_ratio_per_token=log_ratio`. Renamed the KL local to `kl_log_ratio`:

```python
kl_log_ratio = ref_per_token_logps - per_token_logps
if self.args.kl_log_ratio_clip is not None:
    kl_log_ratio = kl_log_ratio.clamp(-self.args.kl_log_ratio_clip, self.args.kl_log_ratio_clip)
per_token_kl = torch.exp(kl_log_ratio) - kl_log_ratio - 1
```

Checked the duplicated K3 blocks: `gmpo` had the same shadowing (its `log_ratio` feeds the clip-fraction metrics), so renamed there too; `gspo_token` was safe by ordering but renamed for consistency. cc @qgallouedec

**Comment by behroozazarkhalili** [comment]
Fair point, and I would rather not add public config surface you have to keep stable if the KL path is winding down.

The crash itself is still real for anyone running `beta != 0` (the K3 `exp(ref - cur)` overflows to `inf` on entropy collapse), but it can be fixed without a new parameter. A few lines up, the sequence-importance path already clamps a log-ratio unconditionally:

```python
# grpo_trainer.py, existing
log_ratio_clamped = torch.clamp(log_ratio_per_token, -20.0, 20.0)
```

So I can drop `kl_log_ratio_clip` and instead clamp the K3 log-ratio to a fixed `[-20, 20]`, consistent with the above. `exp(20)` is about `4.8e8`, so it is a no-op for any legitimate per-token KL and only bites in the overflow regime. That fixes the crash, adds no new public API, and keeps the three duplicated K3 blocks (`grpo`, `gmpo`, `gspo_token`) consistent.

Want me to switch #6637 to that (unconditional clamp, no new field)? If you would rather not touch the KL path at all given its declining use, I am equally happy to just close this. cc @albertvillanova

**Comment by behroozazarkhalili** [comment]
Follow-up: Bugbot just flagged that `kl_log_ratio_clip` is silently ignored on the Liger path. The fused `LigerFusedLinearGRPOLoss` computes the K3 KL inside the kernel, so a Python-side clip never runs, and there is no init-time guard for the combination.

That is actually another reason to prefer the unconditional-clamp route: with no public `kl_log_ratio_clip` field there is no option that can silently no-op on Liger. The non-Liger path still gets the overflow fix, and the Liger kernel's internal K3 stays a separate upstream concern. So dropping the parameter resolves both your config-surface point and the Liger finding in one go. Happy to push that version whenever you give the nod.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -3090,9 +3090,12 @@ def _compute_loss(self, model, inputs):
...
             ref_per_token_logps = inputs["ref_per_token_logps"]
-            per_token_kl = (
-                torch.exp(ref_per_token_logps - per_token_logps) - (ref_per_token_logps - per_token_logps) - 1
-            )
+            kl_log_ratio = ref_per_token_logps - per_token_logps
+            # Clip the log-ratio before the exponential so the K3 estimator stays finite when the policy and
+            # reference distributions drift far apart (issue #3015). No-op when kl_log_ratio_clip is None.
+            if self.args.kl_log_ratio_clip is not None:
+                kl_log_ratio = kl_log_ratio.clamp(-self.args.kl_log_ratio_clip, self.args.kl_log_ratio_clip)
+            per_token_kl = torch.exp(kl_log_ratio) - kl_log_ratio - 1
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:3216]: Good catch, fixed in 4869d491. The Liger fused path computes the KL penalty internally and can't receive `kl_log_ratio_clip`, so rather than silently ignore the clip it now raises `NotImplementedError` at init when `use_liger_kernel=True` with `beta != 0` and `kl_log_ratio_clip` set, consistent with the other Liger-incompatible-option guards. Added a `require_liger_kernel` test asserting the raise. cc @qgallouedec

**Comment by behroozazarkhalili** [comment]
Verification on GPU, since the Liger path cannot be exercised on a CPU runner.

Ran on one H100 at the pushed head `390356b9`, torch 2.11.0 with triton 3.6.0, in two passes.

The 13 Liger nodes: 13 passed, exit 0, in 232s. These are the tests that go red locally without a GPU, where triton raises "0 active drivers" before any assertion runs, so a red there says nothing about the code. On real hardware they pass.

The four `kl_log_ratio_clip` cases (5 selected): 5 passed, exit 0, in 53s. That covers the clip staying finite above the dtype's `exp` threshold and the one-sided clamp leaving a large negative log-ratio alone.

This is the same hardware question raised on the Bugbot thread about the Liger path. The fused loss still raises `NotImplementedError` when `kl_log_ratio_clip` is set, and that guard is among the 13.

**Comment by qgallouedec** [comment]
were you able to reproduce the described behaviour? I don't know what setting would make the ref_logp and the policy_logp drift so far apart. Also I don't think anyone trains with beta!=0 anymore. But maybe I miss some context?

**Comment by behroozazarkhalili** [comment]
Not in a real run. The overflow in the PR is reproduced on CPU with hand-built `per_token_logps` and `ref_per_token_logps` passed straight into `_compute_loss`; float32 `exp` overflows once the log-ratio passes 88.72, so a single token has to be about e^88 less likely under the policy than under the reference. I do not have a training setting that produces that, and I did not try to build one. The only reports of it happening are the two on #3015 (Boltzmachine and Alex-HaochenLi, both in 2025, when `beta` still defaulted to 0.04), and neither posted a config.

On `beta`: agreed that the default is 0.0 now and the reference model is not even loaded then, so this branch is dead for the default path and only reachable for users who opt into KL. That narrows the audience to the two reporters and whoever still follows the DeepSeek-R1 value of 0.001.

The PR exists because @albertvillanova asked for approach (a) on #3015 in June, not because I hit the failure. If you would rather not carry an opt-in guard for a case with no reproducible setting, I am fine closing this and leaving #3015 to the reporters to come back with a config.
